Week3 [STEP 1] MadCow

madcow95 opened the pull request:
안녕하세요 @Allie-e! 벌써 절반인 3주차네요 ㅎㅎ
Week3 첫번째 과제 PR 보냅니다!

# 신경쓴 점
- class와 struct의 차이점을 공부해보며 Person은 struct, CoffeeShop은 class로 만들었습니다.
- 타입에 정의된 함수나 프로퍼티를 사용할 때 타입 내부에 존재한다는 의미로 self를 붙였습니다.
- coffee 열거형에서 case에 따른 이름과 가격도 확인할 수 있도록 만들었습니다.
- 열거형을 사용할 때 CaseIterable protocol을 채택해 allCases를 사용해서 모든 case를 확인할 수 있도록 만들었습니다.
- "돈이라는 속성을 가질 수 있도록 해봅시다."에서 기본값으로 150,000을 가지고 있도록 설정했습니다.
- "매출액을 속성으로 가질 수 있도록 해봅시다."에서 매출액은 커피 주문이 들어왔을 때 커피 가격을 더하도록 구현했습니다.

# 해결이 되지 않은 점
- 과제 중 "주문을 받고, 커피를 만들어낼 수 있는 동작을 가질 수 있도록 해봅시다." 라는 내용이 있는데 "커피를 만들어낸다."의 의미가 어떤 뜻인지 이해가 되지 않아 구현하지 못했습니다ㅠㅠ

# 조언을 얻고 싶은 부분
- 과제 "메뉴판(커피 종류, 가격)을 가질 수 있도록 해봅시다."를 진행하면서 하드코딩으로 만들고 싶지 않아 class내 함수를 통해 메뉴판을 만들기로 했습니다.
-- CoffeeShop class에서 coffeeMenu에 인스턴스가 생성될 때(init()) loadCoffeeMenu() 함수를 실행시켜 coffee enum의 커피들을 저장했습니다.
```swift
class coffeeShop {
    private var coffeeMenu: [String: String] = [:]
    
    init() {
        self.loadCoffeeMenu()
    }

    private func loadCoffeeMenu() {
        Coffee.allCases.forEach{ coffee in
            self.coffeeMenu[coffee.coffeeName] = "\(coffee.coffeePrice)원"
        }
    }

    func showCoffeeMenu() {
        print(self.coffeeMenu)
    }
}

let coffeeShop = coffeeShop()
coffeeShop.showCoffeeMenu()
```
이 방법을 인스턴스에서 프로퍼티에 접근할 때? 커피 메뉴를 불러오는 방식으로 표현해 봤는데 어떤 방식으로 사용하는게 더 일반적인? 방법인지 궁금합니다!
```swift
class coffeeShop {
    lazy var coffeeMenu: [String: String] = self.loadCoffeeMenu()
    
    private func loadCoffeeMenu() -> [String: String] {
        var coffeeMenus: [String: String] = [:]
        Coffee.allCases.forEach{ coffee in
            coffeeMenus[coffee.coffeeName] = "\(coffee.coffeePrice)원"
        }
        return coffeeMenus
    }
}

let coffeeShop = coffeeShop()
print(coffeeShop.coffeeMenu)
```
- 이번 과제를 진행하며 for 와 forEach를 번갈아가며 사용했는데 둘의 속도를 측정해보니 큰 차이가 없었습니다. map과 filter는 사용법을 어느정도 명확하게 알겠는데 for와 forEach는 어떤 경우에 구분해서 사용하는지 궁금합니다!
- coffee 열거형에서 coffeeName, coffeePrice 이렇게 변수명을 설정했는데, coffee라는 이름이 중복되니 coffee를 지우고 name, price로 설정을 했어도 괜찮았을까요??

madcow95 committed:
WEEK3 STEP1

madcow95 committed:
화면 표시 method 추가

madcow95 committed:
Person class -> struct type으로 수정

madcow95 committed:
struct 접근제어 추가

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,143 @@
+  - 사람이 공통적으로 할 수 있는 동작을 메서드로 정의해봅시다.
+    커피를 구매할 수 있도록 메서드를 정의해봅시다.
+ */
+struct Person {
```
Person은 구조체로, CoffeeShop은 클래스로 구현해주셨는데 각각의 이유가 있을까요?

madcow95 replied:
큰 이유는 아니었습니다 ㅎㅎ struct와 class를 한 곳에 쓰면서 차이점을 비교해보려고 사용했습니다!

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,143 @@
+    커피를 구매할 수 있도록 메서드를 정의해봅시다.
+ */
+struct Person {
+    var name: String = ""
+    var gender: Gender = Gender.man
+    var MBTI: String = ""
+    var age: Int = 0
```
프로퍼티에 기본값을 할당하고 있는데, 이니셜라이저를 사용한다면 어떻게 작성할 수 있을까요? 👀

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,143 @@
+        case woman
+    }
+    
+    mutating func buyCoffee(menu: Coffee) {
+        print("\(menu.coffeeName)를 구매했습니다!")
+        self.currentMoney -= menu.coffeePrice
```
currentMoney에 접근제어를 붙여 외부에서 접근할 수 없도록하고,
변화가 있을 때 메서드로 처리한 부분 좋습니다 💯

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,143 @@
+        self.currentMoney -= menu.coffeePrice
+    }
+    
+    func selfIntroduce() {
+        print("제 이름은 \(self.name)구요. 나이는 \(self.age)살, MBTI는 \(self.MBTI)입니다.")
+    }
+    
+    func walk() {
+        print("걷는다")
+    }
+    
+    mutating func spendMoney(price: Int) {
+        self.currentMoney -= price
+    }
+    
+    mutating func makeMoney(salary: Int) {
+        self.currentMoney += salary
+    }
+    
+    func showCurrentMoney() {
+        print("현재 잔액: \(self.currentMoney)")
+    }
```
mutating 키워드가 붙은 메서드와 그렇지 않은 메서드가 번갈아가며 작성되어있네요.
코드 순서를 일정 기준으로 통일해보는건 어떠세요?ㅎㅎ

madcow95 replied:
말씀을 듣고 정리해 보니 훨씬 보기가 편하네요! 이런 점은 왜 항상 나중에 보이는지ㅠㅠ

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,143 @@
+    private var pickUpTable: [Coffee] = []
+    var barista: Person = Person()
+    
+    init() {
+        self.loadCoffeeMenu()
+    }
```
클래스에서 이니셜라이저를 꼭 사용해야하는 이유가 무엇일까요 👀?

madcow95 replied:
CoffeeShop 클래스에서 커피 메뉴를 보여줄 때 enum에 따라 보여주고 싶었는데, 직접 입력하는 것보다는 인스턴스를 생성할 때 coffeeMenu 딕셔너리 변수에 저장하는 함수를 만들어서 보여주는 방법을 사용해 보았습니다!
위와 같은 기능을 구현할 때 이니셜라이저 밖에 떠오르지 않아 사용하게 되었는데 혹시 다른 방법이 있을까요??

madcow95 replied:
```swift
func getCoffeeMenus() -> [String: String] {
    var coffeeMenus: [String: String] = [:]
    Coffee.allCases.forEach{ coffee in
        coffeeMenus[coffee.coffeeName] = "\(coffee.coffeePrice)원"
    }
    return coffeeMenus
}
let coffeeMenus = getCoffeeMenus()
let yagombucks = CoffeeShop(coffeeShopName: "야곰 커피샵", totalSales: 0, coffeeMenu: coffeeMenus, pickUpTable: [], barista: misterLee)
```
커피 메뉴를 생성하는 함수를 외부로 빼서 인스턴스를 생성할 때 넣는 방법으로 수정해봤습니다!

Allie-e replied:
앗 제 의도는 이니셜라이저를 사용하면 안된다는 것이 아니라 구조체와는 다르게 클래스에서는 이니셜라이저를 구현해야하는 이유를 알고계신지 물어보기 위함이었습니다 😊

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,143 @@
+        self.loadCoffeeMenu()
+    }
+    
+    func order(coffee menu: Coffee) {
+        self.pickUpTable.append(menu)
+        self.totalSales += menu.coffeePrice
```
네이밍만 봤을 때는 손님이 커피를 주문하는 것 같아서 조금 어색한 것 같아요!

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,5 +8,143 @@
+}
+
+// 4. Person 타입의 인스턴스로 misterLee , missKim 을 생성해봅시다.
+var misterLee = Person()
+misterLee.name = "MadCow"
+misterLee.MBTI = "ISTP"
+misterLee.age = 28
```
misterLee를 만들면서 한번에 이름, MBTI, 나이를 할당할 수 있는 방법은 없나요?

madcow95 replied:
클래스의 이니셜라이저를 사용하여 인스턴스를 생성할 때 인자를 전달하여 할당할 수 있습니다!
인스턴스를 생성하는 동시에 할당하는 방법이 더 일반적인가요??

Allie-e replied:
이부분도 방법의 차이일뿐 어느 쪽이 더 일반적이다라고 할 수 없는 부분인 것 같아요!

Allie-e reviewed (commented):
광우님, 이번 스텝도 고생하셨습니다 ㅎㅎ
몇가지 코멘트 달아드렸으니 확인해주세요~
추가로 commit message 규칙에 대해 알아보시고 앞으로 일정한 규칙과 양식으로
커밋메세지를 작성해보시는 것을 추천드립니다 😊 (필수는 아니에요!)

### 해결되지 않은 부분 및 질문 주신 부분에 대해
1. 커피를 만들어 낸다 -> 카페에서 커피를 만들고 나면 무엇을 할까요? 예를들면, 고객을 부를수도있고 커피를 픽업대에 올려놓을 수도 있겠죠? `func brew(coffee: Coffee)` 이런식으로 간단하게 만들어 볼 수 있지 않을까요? ㅎㅎ 내부에 print를 하셔도 되고 자유롭게 구현하시면 될 것 같습니다 😃
2. 지금 코드에서는 loadCoffeeMenu 와 showCoffeeMenu로 구분되어있어서 함수의 역할을 분리하는 측면에서는 명확해서 좋다는 생각도 들어요. 현재 방법도 일리가 있다고 생각하는데 어떤 포인트에서 질문해주신건지 잘 와닿지가 않습니다ㅜㅜ! (많이 쓰이는 방법인지를 물으신거라면 항상 일반적인게 정답은 아니에요!)
3. **for-in, forEach**
가장 큰 차이점은 for문은 직접 구현하는 반복문이고, forEach는 반복하고 싶은 부분을 클로저로 넘겨준다는 점인 것 같아요. 클로저라는 개념에 대해서 좀 더 찾아보시고 공부해보시면 이해하는데 도움이 될 것 같습니다!
4. 사용할 때 Coffee.name, Coffee.price로 접근하게 되니 반복되는 부분을 줄여도 좋겠네요 👍🏻

madcow95 committed:
1. 인스턴스 생성 후 각 변수에 할당 -> 인스턴스 생성과 동시에 할당으로 수정
2. 함수 및  변수 키워드 유무에 따라 정렬
3. 뜻이 모호한 함수 이름 수정(func order(coffee menu: Coffee) -> func receivedOrder(coffee menu: Coffee))
4. CoffeeShop Class 바리스타 교체 기능 추가

madcow95 commented:
@Allie-e 코멘트 달아주신 내용을 보고 수정해봤습니다! 좀 더 기능을 추가하려 했지만 STEP2의 과제와 겹치는 부분이 많아 일부분만 수정했습니다 ㅎㅎ.

Allie-e commented:
고생하셨습니다 👍🏻
